r3bdata/califaData/R3BCalifaClusterData.h

```diff
@@ -21,82 +21,61 @@
-     *@param fNbOfCrystalHits		Crystal unique identifier
      *@param fEnergy				Total energy deposited
      *@param fTheta					Reconstructed theta
      *@param fPhi					Reconstructed phi
      **/
-    R3BCalifaClusterData(UInt_t Nb, Double_t ene, Double_t nf, Double_t ns, Double_t theta, Double_t phi, ULong64_t time);
-
-    R3BCalifaClusterData(uint64_t time, double theta, double phi, uint32_t clusterId)
-        : fTime(time)
-        , fTheta(theta)
-        , fPhi(phi)
-        , fClusterId(clusterId)
-        , fEnergy(0.)
-        , fNf(0.)
-        , fNs(0)
-        , fNbOfCrystalHits(0)
-    {
-    }
+    R3BCalifaClusterData(std::vector<Int_t> crystalList, Double_t ene, Double_t nf, Double_t ns, Double_t theta, Double_t phi, ULong64_t time,Int_t clusterType);
 
     /** Copy constructor **/
     R3BCalifaClusterData(const R3BCalifaClusterData&);
 
-    /** += operator **/
-    R3BCalifaClusterData& operator+=(R3BCalifaCrystalCalData& cH)
-    {
-        this->fEnergy += cH.GetEnergy();
-        this->fNf += cH.GetNf();
-        this->fNs += cH.GetNs();
-        this->fNbOfCrystalHits++;
-        return *this;
-    }
 
     R3BCalifaClusterData& operator=(const R3BCalifaClusterData&) { return *this; }
 
     /** Destructor **/
     virtual ~R3BCalifaClusterData() {}
 
     /** Accessors **/
-    UInt_t GetNbOfCrystalHits() const { return fNbOfCrystalHits; }
     Double_t GetEnergy() const { return fEnergy; }
     Double_t GetNf() const { return fNf; }
     Double_t GetNs() const { return fNs; }
     Double_t GetTheta() const { return fTheta; }
     Double_t GetPhi() const { return fPhi; }
     ULong64_t GetTime() const { return fTime; }
-    uint32_t GetClusterId() const { return fClusterId; }
+    uint32_t GetClusterType() const { return fClusterType; }
+    Int_t GetNbOfCrystalHits()  const { return fCrystalList.size(); }
+
+    std::vector<Int_t> GetCrystalList() const {return fCrystalList; }
+
 
     /** Modifiers **/
-    void SetNbOfCrystalHits(UInt_t number) { fNbOfCrystalHits = number; }
     void SetEnergy(Double_t ene) { fEnergy = ene; }
     void SetNf(Double_t nf) { fNf = nf; }
     void SetNs(Double_t ns) { fNs = ns; }
     void SetTheta(Double_t theta) { fTheta = theta; }
     void SetPhi(Double_t phi) { fPhi = phi; }
     void SetTime(ULong64_t time) { fTime = time; }
-    void SetClusterId(uint32_t id) { fClusterId = id; }
-
+    void SetClusterType(uint32_t id) { fClusterType = id; }
+    void SetCrystalList(std::vector<Int_t> list) {fCrystalList = std::move(list);}
   protected:
-    // Basic Hit information
-    UInt_t fNbOfCrystalHits; // number of crystals contribuying to the R3BCalifaClusterData
+
     Double_t fEnergy;        // total energy deposited
     Double_t fNf;            // total Nf deposited
     Double_t fNs;            // total Ns deposited
     Double_t fTheta;         // reconstructed theta
     Double_t fPhi;           // reconstructed phi
     ULong64_t fTime;         // WR time stamp
```

Comment:
Opposed. A cluster is not a CrystalCal object. In particular, even if you name its seed crystalID a clusters fCrystalId, the relationship between crystalId and object is still different for CrystalCal and Cluster. And what about fTotEnergy?

If Clusters were CrystalCal, the Liskov substitution principle would demand that we can operate the clustering with a list of Clusters as input, which is not how things work presently. A case could perhaps be made for replacing CrystalCal with clusters of one entry. Then during clustering, we would duplicate the set of clusters and then merge nearby ones. 

Or we could go overboard and try
```
template <typename T, T def>
struct califaBase
{
   T fEn{def}, fNf{def}, fNs{def};
   uint64_t fWRTS{};
   operator=(...);
   operator+=(...);
};
and derive CrystalCal and Cluster from ``califaBase<double, NAN>`` or something. 

NB: the ULong64_t just made me write [a complaint](https://github.com/root-project/root/issues/12208) about RTypesCore. I figured I am more likely to convince the ROOT devs than my colleagues.

Reply:
Hi Ryo. 
 
> As this is now called Cluster class, it's better to be careful to call any variables with "hit"
That's true, some of them escape to the CalifaHit purge. I'll change any "HIT" to "CLUSTER" 

I think the clustering is clear as it is. There some cases in which you have to be careful adding crystals to a cluster and that is better reflected by using three separated parts instead of a hidden , compact function. But that's only my opinion.  A saturated cluster is a group  in which one or several crystals had an overflow . This allows the user to identify a proton hit, even if you cannot reconstruct the energy. Just imagine a typical (p,2p) event. You have an actual proton cluster on one side of califa and a saturated cluster in the other side. Then it seems reasonable that you can still reconstruct an opening angle or a correlation, or count them for cross sections. 
 

> Also, I think making a time condition (a few ns) is important when applying clustering, otherwise, you will add non-correlated energies.

I thought this was clear, but in Califa we do not have timing in the same way as in Xball, for example. We can correlate events with the main trigger but not internally, so forget about grouping crystals by time.

Reply:
Thanks for the comments. The proposal for using a common data class is just a kind of brainstorming. I am not insisting so I think it is fine to keep as it is so far. But yes, I think mult=1 cluster would be one option actually I think. Then you can define a +operator for adding energy of rhs to lhs, then deleting (or setting a flag to be used) in rhs.

I am not sure about the system of CALIFA timing, but if it can be included that would be great. If it takes time, we can open an issue post for the later development.